### DISCUSSION TITLE
Is there a way to use disable persistent storage for only one query?

### DISCUSSION THREAD

**Discussion by javierguzman** [TanStack/query#4833] [Q&A] [answered]
Hello all,

I am making a react native app where I have enabled the mechanism to save the data on persistent storage so when I init again the app, it loads faster using data from the cache while fetching new stuff. I have followed the docs on:
https://tanstack.com/query/v4/docs/react/plugins/persistQueryClient

However, there is a query I have for fetching avatars and I think it provokes the following warnings:
```
 WARN  [Error: Row too big to fit into CursorWindow requiredPos=0, totalRows=1]
 WARN  Encountered an error attempting to restore client cache from persisted location. As a precaution, the persisted cache will be discarded.
```

Is there a way to tell react-query not to save/load this data? I am open to other solutions, maybe I can extend the storage somehow? I really feel that storing images on the async storage should not be a big deal. I mean Whatsapp for example consumes a lot of data in order to save conversations, attachments, etc.

Thank you in advance and regards

**Comment by TkDodo** [comment, marked as answer]
- you can use retries, either the predefined strategy that we have or with your own strategy: https://tanstack.com/query/v4/docs/react/plugins/createSyncStoragePersister#retries
- you can pass `dehydrateOptions` to `PersistQueryClientProvider` to tell it to exclude a certain query: https://tanstack.com/query/v4/docs/react/plugins/persistQueryClient#persistqueryclient

**Reply by javierguzman** [reply]
Thank you @TkDodo ! I have tried both and they work ok. However, I am still puzzled about this error as in reality I was just storing a few avatars, 3 or 4, as I am in development mode yet. I would really like to save the avatars in async storage so I think I will go for the retries method. One last question, if I use the retry with removeOldestQuery, essentially, when I hit the error I got, react-query will remove the oldest query in order to allocate the avatar query, but this might provoke having to fetch the data removed, is this right?

**Reply by TkDodo** [reply]
> However, I am still puzzled about this error as in reality I was just storing a few avatars, 3 or 4, as I am in development mode yet.

not sure how you store images in the queryCache but apparently they are quite large ...

> One last question, if I use the retry with removeOldestQuery, essentially, when I hit the error I got, react-query will remove the oldest query in order to allocate the avatar query, but this might provoke having to fetch the data removed, is this right?

It will only remove the oldest query from the persistence attempt, and then try to persist again without it until it succeeds. So it will not actually remove it from the cache.

**Reply by CavalcanteLeo** [reply]
Seems not right to me.

```ts
  const pluginOptions: VueQueryPluginOptions = {
    queryClient,
    clientPersister: (queryClient) => {
      return persistQueryClient({
        queryClient,
        persister: createIDBPersister(),
        dehydrateOptions: {
          shouldDehydrateQuery: (query) => {

            const [key1] = query.queryKey

            if (key1 === 'search') { // Open Close principle cries here lol
              return false
            }

            return true
          },
        },
      })
    },
  }
  ```
  
  let's suppose I have 500 endpoints, which about 200 I don't want to cache, do I need to register all keys like this?
  
  isn't it easier if I could pass on any request something like:
  
  
  ```ts
return useQuery<SearchResult>(
    ['search', unref(term)],
    () => {
      return $fetch(url)
    },
    {
      cache: false // <= cache false here would be much easier
      staleTime: 0,
      cacheTime: 0,
    }
  )
}
```

**Reply by TkDodo** [reply]
You can use the meta field for that:

```
return useQuery<SearchResult>({
  queryKey: ['search', unref(term)],
  queryFn: () => $fetch(url),
  meta: {
    cache: false
  }
})
```

and then implement `shouldDehydrateQuery` accordingly:

```
shouldDehydrateQuery: (query) => {
  return query.meta.cache !== false
}
```

be advised that you should probably add the default check as well to only include successful queries:

```
import { defaultShouldDehydrateQuery } from '@tanstack/react-query'

shouldDehydrateQuery: (query) => {
  return defaultShouldDehydrateQuery(query) && query.meta.cache !== false
}
```

**Reply by CavalcanteLeo** [reply]
AMAZING, thanks

**Reply by kyusungpark** [reply]
@TkDodo, would it be possible to have query level maxAge and also keep a default maxAge for the persistor client?

**Reply by lili21** [reply]
@TkDodo can I set meta base on the response data?

**Reply by lili21** [reply]
I'm trying to do it with function. 

```js
useQuery({
  meta: { cacheFn: (data) => !data.xxx }
})

shouldDehydrateQuery = (query: Query) =>
  defaultShouldDehydrateQuery(query) &&
  (query.meta?.cache === true || query.meta?.cacheFn?.(query.state.data));
```

But it doesn't work. it will throw an error

```js
DataCloneError: Failed to execute 'put' on 'IDBObjectStore': function cacheFn(data) {
        return !data.project.is_auto_eval;
      }
```

**Reply by rafcontreras** [reply]
@lili21, make sure the `persistClient` is serialisable:
```ts
const createIDBPersister = ({ key }: { key: string }) =>
  ({
    persistClient: async (client: PersistedClient) => {
      /**
       * Ensure the client is serializable before storing,
       * this is required because we are not storing all
       * of the queries on IndexedDB.
       */
      const serializedClient = JSON.parse(JSON.stringify(client));
      await set(key, serializedClient);
    },
    restoreClient: async () => {
      const client = await get<PersistedClient>(key);
      return client;
    },
    removeClient: async () => {
      await del(key);
    },
  }) as Persister;
```
